cs/src/core/Index/FasterLog/FasterLogIterator.cs

```diff
@@ -0,0 +1,314 @@
+        private readonly long endAddress;
+        private readonly BlittableFrame frame;
+        private readonly CountdownEvent[] loaded;
+        private readonly long[] loadedPage;
+        private readonly LightEpoch epoch;
+        private readonly GetMemory getMemory;
+        private long currentAddress, nextAddress;
+        
+
+        /// <summary>
+        /// Current address
+        /// </summary>
+        public long CurrentAddress => currentAddress;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="fasterLog"></param>
+        /// <param name="hlog"></param>
+        /// <param name="beginAddress"></param>
+        /// <param name="endAddress"></param>
+        /// <param name="scanBufferingMode"></param>
+        /// <param name="epoch"></param>
+        /// <param name="getMemory"></param>
+        internal unsafe FasterLogScanIterator(FasterLog fasterLog, BlittableAllocator<Empty, byte> hlog, long beginAddress, long endAddress, GetMemory getMemory, ScanBufferingMode scanBufferingMode, LightEpoch epoch)
+        {
+            this.fasterLog = fasterLog;
+            this.allocator = hlog;
+            this.getMemory = getMemory;
+            this.epoch = epoch;
+
+            if (beginAddress == 0)
+                beginAddress = hlog.GetFirstValidLogicalAddress(0);
+
+            this.endAddress = endAddress;
+            currentAddress = -1;
+            nextAddress = beginAddress;
+
+            if (scanBufferingMode == ScanBufferingMode.SinglePageBuffering)
+                frameSize = 1;
+            else if (scanBufferingMode == ScanBufferingMode.DoublePageBuffering)
+                frameSize = 2;
+            else if (scanBufferingMode == ScanBufferingMode.NoBuffering)
+            {
+                frameSize = 0;
+                return;
+            }
+
+            frame = new BlittableFrame(frameSize, hlog.PageSize, hlog.GetDeviceSectorSize());
+            loaded = new CountdownEvent[frameSize];
+            loadedPage = new long[frameSize];
+            for (int i = 0; i < frameSize; i++)
+                loadedPage[i] = -1;
+
+        }
+
+        /// <summary>
+        /// Wait for iteration to be ready to continue
+        /// </summary>
+        /// <returns></returns>
+        public async ValueTask WaitAsync()
+        {
+            while (true)
+            {
+                var commitTask = fasterLog.CommitTask;
+                if (nextAddress >= fasterLog.CommittedUntilAddress)
+                    await commitTask;
+                else
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Get next record in iterator
+        /// </summary>
+        /// <param name="entry">Copy of entry, if found</param>
+        /// <param name="entryLength">Actual length of entry</param>
+        /// <returns></returns>
+        public unsafe bool GetNext(out Span<byte> entry, out int entryLength)
```

Comment:
I'm curious if you considered just trimming the `entry` to the precise length instead of returning `entryLength`.
But I'm not sure i fully understand how the buffer return semantics work so maybe that isn't ok

Reply:
The reason I do not trim is that we get the Span from the user via a GetMemory(int minSize) delegate, which may reuse a span that is larger than the requested size. If we returned a truncated span, the caller cannot infer the original size of the span, to be able to reuse in future. Hence, we return the original span and let the user read the first entryLength bytes.

It is a bit convoluted - so am open to other ideas.